libcxx/include/__random/uniform_real_distribution.h

```diff
@@ -27,6 +28,8 @@ _LIBCPP_BEGIN_NAMESPACE_STD
 template<class _RealType = double>
 class _LIBCPP_TEMPLATE_VIS uniform_real_distribution
 {
+  static_assert(is_floating_point<_RealType>::value, "result_type must be floating type");
```

Comment:
We have a utility called `__libcpp_random_is_valid_urng` in [libcxx/include/__random/is_valid.h](https://github.com/llvm/llvm-project/blob/main/libcxx/include/__random/is_valid.h#L57-L61) -- should we instead create something similar to it and handle http://eel.is/c++draft/rand.req.genl more generally?

It seems like the restriction for `is_floating_point` is shared by many distributions. From looking at the synopsis I can find:

```
template<class RealType = double>
class uniform_real_distribution;

template<class RealType = double>
class exponential_distribution;

template<class RealType = double>
class gamma_distribution;

template<class RealType = double>
class weibull_distribution;

template<class RealType = double>
class extreme_value_distribution;

template<class RealType = double>
class normal_distribution;

template<class RealType = double>
class lognormal_distribution;

template<class RealType = double>
class chi_squared_distribution;

template<class RealType = double>
class cauchy_distribution;

template<class RealType = double>
class fisher_f_distribution;

template<class RealType = double>
class student_t_distribution;

template<class RealType = double>
class piecewise_constant_distribution;

template<class RealType = double>
class piecewise_linear_distribution;
```

Reply:
@ldionne Ide
are you suggesting to replace `is_floating_point` by `__libcpp_random_is_valid_urng` in static assert?

Reply:
Add the following to `libcxx/include/__random/is_valid.h`:

```
// [rand.req.genl]/1.4:
// The effect of instantiating a template that has a template type parameter
// named RealType is undefined unless the corresponding template argument is
// cv-unqualified and is one of float, double, or long double.

template<class> struct __libcpp_random_is_valid_realtype : false_type {};
template<> struct __libcpp_random_is_valid_realtype<float> : true_type {};
template<> struct __libcpp_random_is_valid_realtype<double> : true_type {};
template<> struct __libcpp_random_is_valid_realtype<long double> : true_type {};
```

Then, in the distributions mentioned above, add the following `static_assert`:

```
static_assert(__libcpp_random_is_valid_realtype<_RealType>::value, "RealType must be a supported floating-point type");
```